File: Exercises/rb101-rb109_small_problems/medium_2/triangle_sides.rb
=begin

# Problem
- input: 3 numerics
- output symbol
- rules:
  - Valid triangle:
    - sum of the lengths of the two shortest sides must be greater than the length
      of the longest side
    - all sides must have lengths > 0
  - equilateral:
    - all sides are equal lengths
  - isosceles:
    - 2 sides are equal lengths
  - scalene:
    - 0 sides are equal lengths
  - input lengths can be an integer or a float

# Algorithm
Approach:
  - Verify if the input lengths create a valid triangle
  - Determine which type of triangle

1. Write a method that verifies if a valid triangle:
  1.1. Initialize a local variable `lengths` to an array with each length as an element.
  1.2. Return `true` if the sum of the two smallest lengths are greater than the
       sum of the greatest length. (Use `min` and `max`).
2. Write a method that determines the type of triangle:
  2.1.

=end

def triangle(side1, side2, side3)
  lengths = [side1, side2, side3]
  return :invalid unless lengths.min(2).sum > lengths.max

  case lengths.uniq.size
  when 1  then :equilateral
  when 2  then :isosceles
  else         :scalene
  end
end

p triangle(3, 3, 3) #== :equilateral
p triangle(3, 3, 1.5) #== :isosceles
p triangle(3, 4, 5) #== :scalene
p triangle(0, 3, 3) #== :invalid
p triangle(3, 1, 1) #== :invalid

File: ls215/exercises/small_problems/medium_2/triangle_sides.js
/* eslint-disable */

/*
### Problem Description ###

A triangle is classified as follows:

Equilateral: All three sides are of equal length.
Isosceles: Two sides are of equal length, while the third is different.
Scalene: All three sides are of different lengths.

To be a valid triangle, the sum of the lengths of the two shortest sides must be greater than the length of the longest side, and every side must have a length greater than 0. If either of these conditions is not satisfied, the triangle is invalid.

Write a function that takes the lengths of the three sides of a triangle as arguments and returns one of the following four strings representing the triangle's classification: 'equilateral', 'isosceles', 'scalene', or 'invalid'.

# Problem
- input: integer (3)
- output: string
- rules:
  - explicit:
    - triangle definitions:
      - equilateral: all sides are equal
      - isosceles: two sides are equal
      - scalene: 0 sides are equal 
    - valid triangle:
      - sum of lengths of the two shorter sides must be greater than the longest
      - length of every side must be greater than 0

# Data Structures
- input: integer
- working with input:
  - array (to hold sides)
- output: string

# Algorithm
- sort array from smallest to longest length;
- validate input for valid triangle; return 'invalid' or continue
- conditional flow:
  - if all sides are equal to one side, return 'equilateral'
  - if the first and second or second and third elements are equal, return 'isosceles'
  - else return 'scalene'

*/

/* eslint-enable */

function triangle(...sides) {
  sides.sort((a, b) => a - b);

  if (!isValidTriangle(sides)) return 'invalid';

  if (sides.every(side => side === sides[0])) {
    return 'equilateral';
  } else if (sides[0] === sides[1] || sides[1] === sides[2]) {
    return 'isosceles';
  } else {
    return 'scalene';
  }
}

function isValidTriangle(sides) {
  return sides[0] + sides[1] > sides[2] &&
    sides.every(side => side > 0);
}

triangle(3, 3, 3);        // "equilateral"
triangle(3, 3, 1.5);      // "isosceles"
triangle(3, 4, 5);        // "scalene"
triangle(0, 3, 3);        // "invalid"
triangle(3, 1, 1);        // "invalid"
